File: Study Plan/Algorithm I/01_Matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m = len(mat)
        n = len(mat[0])
        MAX = 1000000
        for i in range(m):
            for j in range(n):
                if mat[i][j] != 0:
                    if (i - 1) >= 0:
                        top = mat[i - 1][j]
                    else:
                        top = MAX
                    
                    if (j - 1) >= 0:
                        left = mat[i][j - 1]
                    else:
                        left = MAX
                    mat[i][j] = 1 + min(top, left)
        
        for i in range(m - 1 ,-1, -1):
            for j in range(n - 1, -1, -1):
                if mat[i][j] != 0:
                    if (i + 1) < m:
                        bottom = mat[i + 1][j]
                    else:
                        bottom = MAX
                    
                    if (j + 1)< n:
                        right = mat[i][j + 1]
                    else:
                        right = MAX
                    mat[i][j] = min(mat[i][j], 1 + min(bottom, right))
        
        return mat

File: Study Plan/Algorithm I/01_Matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
    int n = mat.size();
    int m = mat[0].size();
    vector<vector<int>> ans(n, vector<int>(m));
    queue<array<int, 3>>q;
    for(int i = 0; i < n; i ++)
    {
        for(int j = 0; j < m; j ++)
        {
            if(mat[i][j] == 0)
            {
                q.push({i, j, 0});
            }
        }
    }
    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};
    while(!q.empty())
    {
        array<int, 3>front = q.front();
        q.pop();
        ans[front[0]][front[1]] = front[2];
        for(int k = 0; k < 4; k ++){
            int nx = front[0] + dx[k];
            int ny = front[1] + dy[k];
            if(nx < 0 || ny < 0 || nx == n || ny == m || mat[nx][ny] != 1)
                continue;
            mat[nx][ny] = 0;
            q.push({nx,ny,front[2] + 1});
        }
    }
    return ans;
    }
};